openerp/addons/base/ir/ir_token.py

```diff
@@ -0,0 +1,103 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+import uuid
+from datetime import timedelta
+from urlparse import urljoin
+
+from odoo import _, api, fields, models
+from odoo.osv import expression
+from odoo.exceptions import AccessError
+
+
+class IrToken(models.TransientModel):
+
+    _name = 'ir.token'
+    _rec_name = 'value'
+
+    @api.model
+    def _get_random_token(self):
+        """Generate a 32 char long pseudo-random string of digits
+
+        UUID4 makes the chance of a collision (unicity constraint) highly
+        unlikely due to the expiration period of 24 hours.
+        """
+        return uuid.uuid4().hex
+
+    value = fields.Char(required=True, default=_get_random_token, size=32)
+    res_model = fields.Char("Related Document Model", required=True,
+        help="Name of the model that will process the token")
+    res_id = fields.Integer("Related Document ID", required=True,
+        help="ID of the record that will process the token")
+    partner_id = fields.Many2one("res.partner", string="Target Contact",
+        help="Optional restricted contact")
+
+    _sql_constraints = [
+        ('token_value_uniq', 'unique(value)', "Duplicated token")
+    ]
+
+    @api.model_cr
+    def _transient_clean_rows_older_than(self, seconds):
+        """Token have a minimum of 24 hours expiration before being cleaned up"""
+        seconds = max(seconds, 24 * 60 * 60)
+        return super(IrToken, self)._transient_clean_rows_older_than(seconds)
+
+    @api.multi
+    def _generate_token_url(self, route):
+        """ Returns a public URL using a custom route
+
+        :param route: URI in the form "/first/second/%s" with mandatory '%s' that
+            will be replaced by the token value.
+        """
+        self.ensure_one()
+
+        if "%s" not in route:
+            raise ValueError(_("Invalid route format %s, missing '%%s' to integrate token)") % route)
+
+        base_url = self.env['ir.config_parameter'].get_param('web.base.url')
```

Comment:
Is web.base.url parameter available to everyone ? I just reviewed a migration that sudo-ize it.